includes/class-newspack-listings-featured.php

```diff
@@ -0,0 +1,536 @@
+ *
+ * @package Newspack_Listings
+ */
+
+namespace Newspack_Listings;
+
+use \Newspack_Listings\Newspack_Listings_Core as Core;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Products class.
+ * Sets up WooCommerce products for listings.
+ */
+final class Newspack_Listings_Featured {
+	/**
+	 * The meta keys for featured listing meta.
+	 */
+	const META_KEYS = [
+		'featured' => 'newspack_listings_featured',
+		'priority' => 'newspack_listings_featured_priority',
+		'query'    => '_newspack_listings_featured_query_priority',
+		'expires'  => 'newspack_listings_featured_expires',
+	];
+
+	/**
+	 * Hook name for the cron job used to check featured expiration dates daily.
+	 */
+	const CRON_HOOK = 'newspack_listings_expiration_checker';
+
+	/**
+	 * Installed version number of the custom table.
+	 */
+	const TABLE_VERSION = '1.0';
+
+	/**
+	 * Option name for the installed version number of the custom table.
+	 */
+	const TABLE_VERSION_OPTION = '_newspack_listings_priority_version';
+
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var $instance
+	 */
+	protected static $instance = null;
+
+	/**
+	 * Main Newspack_Listings_Featured instance.
+	 * Ensures only one instance of Newspack_Listings_Featured is loaded or can be loaded.
+	 *
+	 * @return Newspack_Listings_Featured - Main instance.
+	 */
+	public static function instance() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self();
+		}
+		return self::$instance;
+	}
+
+	/**
+	 * Constructor.
+	 */
+	public function __construct() {
+		register_activation_hook( NEWSPACK_LISTINGS_FILE, [ __CLASS__, 'create_custom_table' ] );
+		add_action( 'plugins_loaded', [ __CLASS__, 'check_update_version' ] );
+		add_action( 'init', [ __CLASS__, 'register_featured_meta' ] );
+		add_action( 'init', [ __CLASS__, 'cron_init' ] );
+		add_action( self::CRON_HOOK, [ $this, 'check_expired_featured_items' ] );
+		add_action( 'save_post', [ __CLASS__, 'set_feature_priority' ], 10, 2 );
+		add_filter( 'posts_clauses', [ __CLASS__, 'sort_featured_listings' ], 10, 2 );
+		add_filter( 'post_class', [ __CLASS__, 'add_featured_classes' ] );
+		add_filter( 'newspack_blocks_term_classes', [ __CLASS__, 'add_featured_classes' ] );
+	}
+
+	/**
+	 * Get events table name.
+	 */
+	public static function get_table_name() {
+		global $wpdb;
+		return $wpdb->prefix . 'newspack_listings_priority';
+	}
+
+	/**
+	 * Checks if the custom table has been created and is up-to-date.
+	 * If not, run the create_custom_table method.
+	 * See: https://codex.wordpress.org/Creating_Tables_with_Plugins
+	 */
+	public static function check_update_version() {
+		$current_version = get_option( self::TABLE_VERSION_OPTION, false );
+
+		if ( self::TABLE_VERSION !== $current_version ) {
+			self::create_custom_table();
+			self::populate_custom_table();
+		}
+	}
+
+	/**
+	 * Create a custom DB table to store feature priority data.
+	 * Avoids the use of slow post meta for query sorting purposes.
```

Comment:
I'm curious about the specific performance gain of the custom table vs. `WP_Query` with meta vs. direct DB query (of the post meta table). Was such testing done?

Reply:
I didn't test a direct DB query of the post meta table, and I'm not sure how I'd quantify the performance in order to compare? The reason we went with a custom table was that it was our understanding that post meta values aren't indexed, so the performance compared to a custom table containing only two columns indexed by the value we want to sort by would always be better than the post meta table could provide—is that an incorrect assumption?